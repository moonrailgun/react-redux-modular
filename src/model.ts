import _get from 'lodash/get';
import _isFunction from 'lodash/isFunction';
import { StandardState } from './types';

type ReduxModelDispatchFn = () => void;

/**
 * A base redux model to be extends
 *
 * class User extends ReduxModel<{
 *    username: string
 * }> {
 *    state = {
 *       username: ''
 *    };
 * }
 */
export abstract class ReduxModel<
  S extends StandardState = { [key: string]: any }
> {
  public dispatch?: ReduxModelDispatchFn;
  protected abstract state: S;

  public getState<K extends keyof S>(key: K): S[K] {
    return _get(this.state, key);
  }

  public getAllState(): S {
    return { ...this.state };
  }

  public setState<K extends keyof S>(state: Pick<S, K> | S | null) {
    this.state = {
      ...this.state,
      ...state,
    };

    if (_isFunction(this.dispatch)) {
      this.dispatch();
    } else {
      console.error('dispatch has not init');
    }
  }
}
