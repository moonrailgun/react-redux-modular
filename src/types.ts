import { ReduxModel } from './model';

export interface StandardState {
  [key: string]: any;
}

export interface ModelMapType {
  [modelName: string]: ReduxModel;
}

export interface StoreStateType {
  [modelName: string]: StandardState;
}

export type StateChangeListenerFn = (
  modelName: string,
  allState: StoreStateType
) => void;

export interface ModelStore {
  models: ModelMapType;
  subscribe: (listener: StateChangeListenerFn) => void;
  getAllState: () => StoreStateType;
}
