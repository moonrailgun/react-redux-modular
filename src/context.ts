import React from 'react';
import { ModelStore } from './types';

// 全局的Context
export const modelContext = React.createContext<ModelStore>({
  models: {},
  // tslint:disable-next-line: no-empty
  subscribe: () => {},
  getAllState: () => ({}),
});
