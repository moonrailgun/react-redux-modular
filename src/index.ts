export { ReduxModel } from './model';
export { Provider } from './Provider';
