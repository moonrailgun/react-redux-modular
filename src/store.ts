import _isFunction from 'lodash/isFunction';
import _mapValues from 'lodash/mapValues';
import { ReduxModel } from './model';
import { ModelStore, StateChangeListenerFn, StoreStateType } from './types';

type ReduxModelCls = new () => ReduxModel;
interface ReduxModelMap {
  [modelName: string]: ReduxModelCls;
}

export const createStore = (initialModels: ReduxModelMap): ModelStore => {
  const models = _mapValues(initialModels, (Model) => new Model());

  const listeners: StateChangeListenerFn[] = [];
  const subscribe = (listener: StateChangeListenerFn) => {
    listeners.push(listener);
  };

  const getAllState = () => _mapValues(models, (model) => model.getAllState());

  Object.entries(models).forEach(([name, model]) => {
    model.dispatch = () => {
      for (const listener of listeners) {
        listener(name, getAllState());
      }
    };
  });

  return {
    models,
    subscribe,
    getAllState,
  };
};
