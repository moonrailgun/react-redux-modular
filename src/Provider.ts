import { createElement, FC } from 'react';
import { modelContext } from './context';

interface ProviderProps {
  store: any;
}
export const ModelProvider: FC<ProviderProps> = (props) => {
  const Provider = modelContext.Provider;

  return createElement(
    Provider,
    {
      value: props.store,
    },
    props.children
  );
};
