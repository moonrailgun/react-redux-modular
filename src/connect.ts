import _mapValues from 'lodash/mapValues';
import {
  ComponentType,
  createElement,
  FC,
  useContext,
  useEffect,
  useMemo,
  useState,
} from 'react';
import { modelContext } from './context';
import { ModelMapType } from './types';

/**
 * connect ReduxModel with modelName
 *
 * Usage:
 * connect(['user', 'counter'])(Component)
 *
 * class Component extends React.Component {
 *  render() {
 *    return <div>{this.props.counter.getState('num')}</div>
 *  }
 * }
 */

export const connectModel = (modelNames: string[]): any => {
  return <P>(Component: ComponentType<P>): FC<P> => {
    return (props) => {
      const { models, subscribe } = useContext(modelContext);
      const modelMap = useMemo(() => {
        const map: ModelMapType = {};
        for (const modelName of modelNames) {
          map[modelName] = models[modelName];
        }

        return map;
      }, [models, modelNames]);

      const [modelState, setModelState] = useState(
        getModelMapAllState(modelMap)
      );
      useEffect(() => {
        subscribe((name) => {
          if (modelNames.includes(name)) {
            setModelState(getModelMapAllState(modelMap));
          }
        });
      }, [models, subscribe]);

      return createElement<P>(Component, {
        ...props,
        modelState, // 这个参数起始可以不传，先传下去方便调试
        ...modelMap,
      });
    };
  };
};

function getModelMapAllState(modelMap: ModelMapType) {
  return _mapValues(modelMap, (model) => model.getAllState());
}
