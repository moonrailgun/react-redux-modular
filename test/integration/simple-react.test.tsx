import { mount, shallow } from 'enzyme';
import React from 'react';
import { connectModel, ModelProvider } from '../../src';
import { createStore } from '../../src/store';
import { CounterModel } from '../models/counter';
import { sleep } from '../utils';

interface CounterProps {
  counter: CounterModel;
}

class Counter extends React.Component<CounterProps> {
  public render() {
    const { counter } = this.props;
    return (
      <div>
        <div id="num">{counter.getState('num')}</div>
        <button onClick={counter.increase}>Increase</button>
        <button onClick={counter.decrease}>Decrease</button>
      </div>
    );
  }
}

const ConnectedCounter = connectModel(['counter'])(Counter);

describe('intergration react component', () => {
  test('render snapshot', () => {
    const store = createStore({
      counter: CounterModel,
    });

    const wrapper = mount(
      <ModelProvider store={store}>
        <ConnectedCounter />
      </ModelProvider>
    );

    expect(wrapper.toJson()).toMatchSnapshot();
  });

  test('store should update', () => {
    const store = createStore({
      counter: CounterModel,
    });

    const wrapper = mount(
      <ModelProvider store={store}>
        <ConnectedCounter />
      </ModelProvider>
    );

    const btns = wrapper.find('div > button');
    expect(btns).toHaveLength(2);

    expect(store.getAllState()).toMatchObject({
      counter: { num: 0 },
    });

    btns.at(0).simulate('click');
    expect(store.getAllState()).toMatchObject({
      counter: { num: 1 },
    });

    btns.at(1).simulate('click');
    expect(store.getAllState()).toMatchObject({
      counter: { num: 0 },
    });
  });

  test.only('dom should update', () => {
    const store = createStore({
      counter: CounterModel,
    });

    const wrapper = mount(
      <ModelProvider store={store}>
        <ConnectedCounter />
      </ModelProvider>
    );

    const btns = wrapper.find('div > button');
    expect(btns).toHaveLength(2);
    const dom = wrapper.find('#num');
    expect(dom).toHaveLength(1);

    expect(dom.text()).toBe('0');

    btns.at(0).simulate('click');
    expect(dom.text()).toBe('1');

    btns.at(1).simulate('click');
    expect(dom.text()).toBe('0');
  });
});
