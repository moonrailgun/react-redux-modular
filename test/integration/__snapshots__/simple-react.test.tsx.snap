// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`intergration react component render snapshot 1`] = `
<ReactModelProvider
  store={
    Object {
      "getAllState": [Function],
      "models": Object {
        "counter": CounterModel {
          "decrease": [Function],
          "dispatch": [Function],
          "increase": [Function],
          "state": Object {
            "num": 0,
          },
        },
      },
      "subscribe": [Function],
    }
  }
>
  <Component>
    <Counter
      counter={
        CounterModel {
          "decrease": [Function],
          "dispatch": [Function],
          "increase": [Function],
          "state": Object {
            "num": 0,
          },
        }
      }
      modelState={
        Object {
          "counter": Object {
            "num": 0,
          },
        }
      }
    >
      <div>
        <div
          id="num"
        >
          0
        </div>
        <button
          onClick={[Function]}
        >
          Increase
        </button>
        <button
          onClick={[Function]}
        >
          Decrease
        </button>
      </div>
    </Counter>
  </Component>
</ReactModelProvider>
`;
